rsocket-core/src/main/java/io/rsocket/internal/UnboundedProcessor.java

```diff
@@ -321,23 +346,29 @@ public void cancel() {
 
   @Override
   public T peek() {
+    if (!priorityQueue.isEmpty()) {
```

Comment:
This looks racy (checking before calling peek) after the change, does Reactor promise this won't be concurrent code in practice?

Reply:
I guess this code is never used in reactor. Thus just a forced Queue API. But indeed, it is racy in I guess i could do nothing to that

Reply:
I meant Reactor in the sense it is a FluxProcessor and Fuseable.QueueSubscription, and are there guarantees within the Reactor framework that indicate the threading model is safe here? It sounds like you are saying it's known safe within our rsocket project.

Reply:
So, by default it is 
```

	interface QueueSubscription<T> extends Queue<T>, Subscription {
		
		String NOT_SUPPORTED_MESSAGE = "Although QueueSubscription extends Queue it is purely internal" +
				" and only guarantees support for poll/clear/size/isEmpty." +
				" Instances shouldn't be used/exposed as Queue outside of Reactor operators.";

		...

		
		@Override
		@Nullable
		default T peek() {
			throw new UnsupportedOperationException(NOT_SUPPORTED_MESSAGE);
		}

```
and as it is written in the error message only poll supposed to be used. So I don't think it matters at all and the best I can do is removing the peek operation at all.

@simonbasle @smaldini can you please correct me if I'm wrong

Reply:
you are correct @OlegDokuka